internal/app/machined/pkg/controllers/etcd/member.go

```diff
@@ -0,0 +1,169 @@
+	return "etcd.MemberController"
+}
+
+const etcdServiceID = "etcd"
+
+// Inputs implements controller.Controller interface.
+func (ctrl *MemberController) Inputs() []controller.Input {
+	return []controller.Input{
+		{
+			Namespace: v1alpha1.NamespaceName,
+			Type:      v1alpha1.ServiceType,
+			ID:        pointer.To(etcdServiceID),
+			Kind:      controller.InputStrong,
+		},
+	}
+}
+
+// Outputs implements controller.Controller interface.
+func (ctrl *MemberController) Outputs() []controller.Output {
+	return []controller.Output{
+		{
+			Type: etcd.MemberType,
+			Kind: controller.OutputExclusive,
+		},
+	}
+}
+
+// Run implements controller.Controller interface.
+//
+//nolint:gocyclo,cyclop
+func (ctrl *MemberController) Run(ctx context.Context, r controller.Runtime, logger *zap.Logger) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-r.EventCh():
+		}
+
+		m := etcd.NewMember(etcd.NamespaceName, etcd.LocalMemberID)
+
+		etcdService, err := safe.ReaderGet[*v1alpha1.Service](ctx, r, v1alpha1.NewService(etcdServiceID).Metadata())
+		if err != nil {
+			if state.IsNotFoundError(err) {
+				continue
+			}
+
+			return fmt.Errorf("error getting etcd service resource: %w", err)
+		}
+
+		switch etcdService.Metadata().Phase() {
+		case resource.PhaseRunning:
+			if err := ctrl.modifyLocalEtcdMember(ctx, r, m, etcdService.TypedSpec().Running); err != nil {
+				return err
+			}
+
+			if err := r.AddFinalizer(ctx, etcdService.Metadata(), ctrl.Name()); err != nil {
+				return fmt.Errorf("error adding finalizer to etcd service resource: %w", err)
+			}
+		case resource.PhaseTearingDown:
+			if err := ctrl.cleanupLocalEtcdMember(ctx, r, m); err != nil {
+				return err
+			}
+
+			if err := r.RemoveFinalizer(ctx, etcdService.Metadata(), ctrl.Name()); err != nil {
+				return fmt.Errorf("error removing finalizer from etcd service resource: %w", err)
+			}
+		}
+	}
+}
+
+func (ctrl *MemberController) cleanupLocalEtcdMember(ctx context.Context, r controller.Runtime, m *etcd.Member) error {
+	etcdMember, err := safe.ReaderGet[*etcd.Member](ctx, r, m.Metadata())
+	if err != nil {
+		if state.IsNotFoundError(err) {
+			return nil
+		}
+
+		return fmt.Errorf("error getting etcd member resource: %w", err)
+	}
+
+	if err := r.Destroy(ctx, etcdMember.Metadata()); err != nil {
+		return fmt.Errorf("error destroying etcd member resource: %w", err)
+	}
+
+	return nil
+}
+
+func (ctrl *MemberController) modifyLocalEtcdMember(ctx context.Context, r controller.Runtime, m *etcd.Member, running bool) error {
+	var memberIDUint64 uint64 = 0
+
+	joining := true
+
+	if running {
+		joining = false
+
+		var err error
+
+		memberIDUint64, err = ctrl.GetLocalMemberID(ctx)
+		if err != nil {
+			return fmt.Errorf("error getting etcd local member ID: %w", err)
+		}
+	}
+
+	memberID := etcd.FormatMemberID(memberIDUint64)
+
+	if err := safe.WriterModify(ctx, r, m, func(status *etcd.Member) error {
+		if memberIDUint64 != 0 {
```

Comment:
not sure what 'joining' means fully, but joining members don't have a member ID, so we should set it to zero if it's joining? or drop joining completely (as it doesn't seem to be used)